Add Python Baseline Solver & Multi-Engine Architecture

jeffersonBastos opened the pull request:
## Overview
This PR introduces the **baseline solver implementation in Python**, ported from the Rust version, and refactors the template into a **multi-engine architecture**.  
The goal is to provide Python developers with:
- A **ready-to-use baseline solver** for comparison and testing,  
- A **`mysolver/solve` endpoint** as a clean starting point for new solver implementations,  
- A modernized, modular, and testable project structure.

---

## Key Changes
### 🆕 New Features
- **Baseline Solver Engine**  
  - Added `BaselineEngine` (`/baseline/solve`) implementing a contract-correct, minimal solver.  
  - Includes core modules: matcher, pathfinding, pool management, solution builder, contract interactions.  

- **Multi-Engine System**  
  - Introduced `SolverEngine` base class for pluggable engines.  
  - Added `mysolver/solve` endpoint (stubbed for user implementations).  

- **New Domain & Utility Modules**  
  - Domain: `auction`, `order`, `solution`, `liquidity`, `eth`, `common`.  
  - Utilities: AMM math, U256 handling, bytes/hex conversion, serialization.  

- **Infrastructure Improvements**  
  - Pydantic v2 models with Rust-aligned DTOs.  
  - Structured logging, Prometheus metrics, and health checks.  
  - CLI entrypoints for local development.  

### 🔄 Refactors & Migrations
- **Single App**: Consolidated legacy server into one FastAPI app.  
- **Routers**: Modular API routing (`/baseline`, `/mysolver`).  
- **Project Structure**: Full reorganization to domain-driven architecture.  
- **Removed Legacy Models & Utils**: Old `src/models` and `src/util` replaced with domain + utils modules.  

### 🗑️ Deletions
- Removed outdated legacy models (batch_auction, order, token, etc.).  
- Removed old test files, replaced with comprehensive new suite.  
- Dropped unused Docker setup (development now CLI-based).  

---

## Motivation
- Provide Python developers with the same onboarding experience Rust solvers already have.  
- Allow **baseline vs custom solver races** locally, without requiring Rust knowledge.  
- Improve maintainability, modularity, and testability for long-term growth.  

---

## Next Steps
- Improve documentation with a **Quick Start guide** on integrating and extending this solver.  
- Add more comprehensive validation tests with complete auctions, similar to Rust’s coverage.  
- Ensure parity validation with the Rust baseline solver.  
- Explore performance improvements.  
- We are open to feedback and suggestions — this is just the first version.

jeffersonBastos committed:
feat: update requirements (pydantic, fastapi and uvicorn)

jeffersonBastos committed:
feat: update README and remove Docker initialization option

jeffersonBastos committed:
fix: update Python solver template for CoW Protocol compatibility

- Fix fundamental schema incompatibility with current CoW infrastructure
- Update models to use snake_case with Pydantic aliases for camelCase JSON
- Replace legacy BatchAuction with CowAuction matching solvers_dto
- Add proper field aliases (sellToken -> sell_token, etc.)
- Remove incompatible legacy models and utilities
- Update server endpoint to use new schema structure
- Organize template structure with examples and reference materials

Resolves schema validation errors (422) when receiving auctions from driver.
Template now works seamlessly with current CoW Protocol infrastructure.

jeffersonBastos committed:
chore: add .tool-versions for development environment setup

jeffersonBastos committed:
chore: update requirements.txt with new dependencies and structure

- Add typer for CLI support
- Add prometheus-client for metrics
- Update version constraints for better compatibility
- Organize dependencies by category (core, dev, optional)

jeffersonBastos committed:
feat: add complete python_baseline module structure

- Add multi-engine solver architecture with baseline and mysolver engines
- Implement FastAPI application with health, metrics, and solve endpoints
- Add comprehensive model definitions for auction, order, liquidity, and solution
- Include utility modules for serialization, math operations, and byte conversion
- Add CLI interface with typer for running the solver server
- Implement dependency injection and settings management
- Add comprehensive test suite for API, models, and multi-engine functionality
- Include logging and metrics infrastructure with Prometheus support

jeffersonBastos committed:
refactor: migrate legacy server to use new baseline structure

- Replace custom FastAPI implementation with python_baseline.api.app
- Remove duplicate endpoint definitions and middleware setup
- Simplify server configuration by delegating to main solver application
- Maintain backward compatibility for existing server startup process

jeffersonBastos committed:
docs: update README with multi-engine solver documentation

- Add comprehensive documentation for new multi-engine architecture
- Include examples for baseline, mysolver, and default engine endpoints
- Update health check endpoint from /health to /healthz
- Add Prometheus metrics endpoint documentation
- Update driver configuration examples with new port (11088)
- Add autopilot configuration examples for different solver engines
- Maintain backward compatibility documentation for legacy server

jeffersonBastos committed:
feat: reorganize project structure

- Move python_baseline/ to src/ with clean structure
- Separate API routers (baseline, mysolver)
- Move models to domain/
- Update all imports from python_baseline.* to src.*
- Remove legacy files and duplicates
- Improve project organization and maintainability

jeffersonBastos committed:
feat: add solver configuration system with pydantic

jeffersonBastos committed:
feat: implement baseline solver with CoW matching and AMM routing

- Add Web3 and blockchain dependencies (web3, eth-typing, eth-utils, eth-abi, hexbytes)
- Add optimization libraries (networkx for pathfinding, scipy for math)
- Implement baseline solver engine with two-phase approach:
  * CoW (Coincidence of Wants) matching for direct order settlement
  * AMM routing through liquidity pools using NetworkX pathfinding
- Add comprehensive AMM math utilities for Uniswap V2, Balancer, and Curve pools
- Create solution builder for transaction construction and settlement
- Add partial fill optimization framework and gas estimation utilities
- Add comprehensive test coverage for baseline engine
- Remove obsolete test files and update domain models
- Fix pydantic field_validator imports and type annotations

jeffersonBastos committed:
fix: add log_format at settings

jeffersonBastos committed:
fix: move validate_auction to shared folder

jeffersonBastos committed:
fix: adjust _generate_solution_id function to follow Rust implementation

jeffersonBastos committed:
chore: rename pathfinder to path_finder

jeffersonBastos committed:
feat: implement AMM integration for baseline solver

jeffersonBastos committed:
fix: change solution id from string to integer

jeffersonBastos committed:
feat: add notify route

jeffersonBastos committed:
fix: update to Pydantic v2 syntax

jeffersonBastos committed:
fix: remove unused functions and file

jeffersonBastos committed:
chore: format code

jeffersonBastos committed:
chore: remove legacy test

jeffersonBastos committed:
feat: create Makefile

jeffersonBastos committed:
fix: adjust tests

jeffersonBastos committed:
chore: update README

yvesfracari commented on .env.example:
remember to delete this file

jeffersonBastos committed:
fix: delete unused files

yvesfracari commented on README.md:
```diff
@@ -1,113 +1,337 @@
-source ./venv/bin/activate
+# Create virtual environment
+python -m venv venv
+source venv/bin/activate
```
Would prefer to use a package manager like `poetry`

jeffersonBastos replied:
Yes, added!

yvesfracari commented on README.md:
```diff
@@ -1,113 +1,337 @@
-- In Depth Solver
-  Specification: https://docs.cow.fi/off-chain-services/in-depth-solver-specification
-- Settlement Contract (namely the settle
-  method): https://github.com/cowprotocol/contracts/blob/ff6fb7cad7787b8d43a6468809cacb799601a10e/src/contracts/GPv2Settlement.sol#L121-L143
-- Interaction Model (Currently missing from this framework): https://github.com/cowprotocol/services/blob/cda5e36db34c55e7bf9eb4ea8b6e36ecb046f2b2/crates/shared/src/http_solver/model.rs#L125-L130
+See LICENSE file for details.
```
why do we need this part here?

jeffersonBastos replied:
For no specific reason, they already had a LICENSE file, I'll remove it.

yvesfracari commented on README.md:
```diff
@@ -1,113 +1,337 @@
-Navigate to [barn.cowswap.exchange/](https://barn.cowswap.exchange/#/swap) and place a
-tiny (real) order. See your driver pick it up and include it in the next auction being
-sent to your solver
+This template is actively maintained. Please feel free to submit issues and pull requests to improve the template for the community.
```
```suggestion
Please feel free to submit issues and pull requests to improve the template for the community.
```

yvesfracari commented on README.md:
```diff
@@ -1,113 +1,337 @@
+
+## Schema Compatibility
+
+This template now uses the current CoW Protocol schema:
```
```suggestion
This template uses the schema:
```

yvesfracari commented on README.md:
```diff
@@ -1,113 +1,337 @@
+
+## Architecture
+
+The solver template now features a clean, modular architecture:
```
Using terms as `now`, `current` is not needed in my opinion. I understand that you added it because it was outdated before, and it is ok to use in PR description, but this can become outdated in the future again 

```suggestion
The solver template features a clean, modular architecture:
```

yvesfracari commented on src/_server.py:
```diff
@@ -37,57 +32,14 @@ class ServerSettings(BaseSettings):
-    print("\n\n*************\n\nReturning solution: " + str(trivial_solution))
-    return trivial_solution
+# Note: The app is now imported from src.api.app
+# All endpoints are handled by the main solver application
```
```suggestion
```

yvesfracari commented on src/api/app.py:
```diff
@@ -0,0 +1,175 @@
+app = FastAPI(
+    title="CoW Protocol Solver",
+    description="Multi-engine solver implementation for CoW Protocol auctions",
+    version="1.0.0",
```
```suggestion
    version="0.0.1",
```

yvesfracari commented on src/domain/eth.py:
did you consider use some web3 package that already have this implemented like `eth_utils` or `web3`?

yvesfracari commented on src/engines/baseline/cow_matcher.py:
```diff
@@ -0,0 +1,266 @@
+
+        # Calculate traded amounts
+        # This is simplified - real implementation needs to handle partial fills
+        traded_amount = min(int(sell_order.sell_amount), int(buy_order.buy_amount))
```
I would add a tag for this and mention it on the README so it is easier to catch all the non production ready parts of the template

jeffersonBastos replied:
Done:
https://github.com/bleu/cowprotocol-solver-template-py/pull/4/commits/dc4e4397058d5350916ec39547f55b5ad49614dd

yvesfracari commented on src/engines/baseline/interactions.py:
```diff
@@ -0,0 +1,270 @@
+except ImportError:
+    # Fallback for environments without web3
+    Web3 = None
+    encode = None
```
why we need this? I think that we need to make a decision if we want or not to use web3

jeffersonBastos replied:
Yes, you're right, it was added as a bug fix and I ended up committing it

yvesfracari commented on src/infra/metrics.py:
This seems to be duplicated with `src/api/metrics.py` or I am missing something?

yvesfracari commented on src/utils/amm_math.py:
did you search for packages that already have this implemented?

I think that balancer math is inside `balpy` and just found this as well:

https://uniswap-python.com/getting-started.html

jeffersonBastos replied:
I took a look, but decided not to use the libs.
1 - To maintain compatibility with the Rust implementation.
2 - The libs are optimized for direct trading, not for solvers.
3 - The calculations seemed simple enough to implement directly.

yvesfracari commented on src/utils/bytes_conv.py:
if we use a web3 package this will not be needed as well

jeffersonBastos committed:
fix: remove web3 and networkx fallback

jeffersonBastos committed:
chore: remove dockerignore file

jeffersonBastos committed:
fix: pydantic_settings import

jeffersonBastos committed:
refactor: consolidate metrics and remove redundant code

jeffersonBastos committed:
feat: add TODO tags for non-production ready parts

jeffersonBastos committed:
feat: migrate from venv/pip to Poetry for dependency management

jeffersonBastos committed:
feat: update github workdlow to use Poetry and create DockerFile

jeffersonBastos committed:
fix: adjust readme and MakeFile

jeffersonBastos committed:
feat: refactor domain models and update solver configurations

- Simplify domain models by removing custom U256/Address types
- Migrate to Pydantic v2 field validators
- Add centralized fee conversion utilities
- Update solver configurations with Python and baseline endpoints
- Expand base token list with popular tokens (WETH, DAI, USDC, USDT, etc.)
- Add Uniswap V2 configuration support
- Integrate Web3 for robust Ethereum address validation
- Remove deprecated utility modules (hexbytes, u256)
- Update interaction encoding to use HexBytes
- Improve type safety with proper address validation

jeffersonBastos committed:
fix: adjust pull-request.yaml
